cpp/src/mcts/main/UnitTests.cpp

```diff
@@ -0,0 +1,233 @@
+
+        oss << std::string(num_indent * 2, ' ') << "|-"
+        << "Edge " << edge_index
+        << ": " << " E = " << edge->N << ", Action = " << edge->action << std::endl;
+        if (edge->child_index != -1) {
+          oss << print_tree(edge->child_index, new_state, num_indent + 2);
+        }
+      }
+      return oss.str();
+    }
+
+    /*
+     * This function prints the tree structure starting from root.
+     * It recursively traverses the tree and prints each node and its edges.
+     * The format is as follows:
+     * - Each node is printed with its index, state representation, RN (visit count), and Q values.
+     * - Each edge is printed with its index, visit count (E), and action.
+     * - Indentation is used to represent the tree structure, with each level of depth indented further.
+    */
+    std::string print_tree() {
+      State state;
+      Game::Rules::init_state(state);
+      return print_tree(0, state);
+    }
+
+ private:
+  ManagerParams manager_params_;
+  Manager manager_;
+};
+
+TEST_F(ManagerTest, construct_tree) {
+  start_manager();
+  start_threads();
+  search(10);
+  EXPECT_EQ(print_tree(),
+  "Node 0: [21, 0] RN = 11: Q= 0.5 0.5\n"
+  "|-Edge 0:  E = 4, Action = 0\n"
+  "  |-Node 1: [20, 1] RN = 4: Q= 0.5 0.5\n"
+  "    |-Edge 3:  E = 1, Action = 0\n"
+  "      |-Node 4: [19, 0] RN = 1: Q= 0.5 0.5\n"
+  "    |-Edge 4:  E = 1, Action = 1\n"
+  "      |-Node 5: [18, 0] RN = 1: Q= 0.5 0.5\n"
+  "    |-Edge 5:  E = 1, Action = 2\n"
+  "      |-Node 6: [17, 0] RN = 1: Q= 0.5 0.5\n"
+  "|-Edge 1:  E = 3, Action = 1\n"
+  "  |-Node 2: [19, 1] RN = 3: Q= 0.5 0.5\n"
+  "    |-Edge 6:  E = 1, Action = 0\n"
+  "      |-Node 5: [18, 0] RN = 1: Q= 0.5 0.5\n"
+  "    |-Edge 7:  E = 1, Action = 1\n"
+  "      |-Node 6: [17, 0] RN = 1: Q= 0.5 0.5\n"
+  "|-Edge 2:  E = 3, Action = 2\n"
+  "  |-Node 3: [18, 1] RN = 3: Q= 0.5 0.5\n"
+  "    |-Edge 9:  E = 1, Action = 0\n"
+  "      |-Node 6: [17, 0] RN = 1: Q= 0.5 0.5\n"
+  "    |-Edge 10:  E = 1, Action = 1\n"
+  "      |-Node 7: [16, 0] RN = 1: Q= 0.5 0.5\n");
+}
+
+TEST_F(ManagerTest, backprop) {
+  start_manager();
+  start_threads();
+  search(0);
+
+  init_search_thread();
+  node_pool_index_t child_index1 = add_child_by_action(0, 1);
+  node_pool_index_t child_index2 = add_child_by_action(child_index1, 2);
+  node_pool_index_t child_index3 = add_child_by_action(child_index2, 0);
+  ValueArray v;
+  v << 0.8, 0.2;
+  modify_node_Q(child_index2, v);
+  modify_node_RN(child_index2, 100);
+  v << 0.1, 0.9;
+  modify_node_Q(child_index3, v);
+  modify_node_RN(child_index3, 9999);
+
+  EXPECT_EQ(print_tree(),
+  "Node 0: [21, 0] RN = 1: Q= 0.5 0.5\n"
+  "|-Edge 1:  E = 0, Action = 1\n"
+  "  |-Node 1: [19, 1] RN = 0: Q= 0.5 0.5\n"
+  "    |-Edge 5:  E = 0, Action = 2\n"
+  "      |-Node 2: [16, 0] RN = 100: Q= 0.8 0.2\n"
+  "        |-Edge 6:  E = 0, Action = 0\n"
+  "          |-Node 3: [15, 1] RN = 9999: Q= 0.1 0.9\n");
+  }
```

Comment:
I don't understand this test. It purports to test `backprop`, but I don't see a `backprop` function getting called?